### PR TITLE
feat(NotificationDrawer): Add new Notification Drawer Component

### PR DESCRIPTION
**Pull request by jcaianirh**
Introduce notification drawer component and notification group.

re #1939

preview: https://patternfly-react-pr-4137.surge.sh/documentation/react/components/notificationdrawer

### CODE REVIEW

File: packages/react-core/src/components/NotificationDrawer/NotificationDrawerGroup.tsx [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        <button
+          className={css(styles.notificationDrawerGroupToggle)}
+          aria-expanded={isExpanded}
+          onClick={e => onExpand(e, !isExpanded)}
+        >
+          <div>{title}</div>
+          <div className={css(styles.notificationDrawerGroupToggleCount)}>
+            <Badge isRead={isRead}>{count}</Badge>
+          </div>
+          <span className="pf-c-notification-drawer__group-toggle-icon">
```
Comment (redallen) [inline packages/react-core/src/components/NotificationDrawer/NotificationDrawerGroup.tsx:51]: If you rebase, I believe react-styles should have a `styles.xyz` class for this.

File: packages/react-integration/cypress/integration/notificationdrawergroups.spec.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      .should('have.class', 'pf-m-danger');
+    cy.get('.pf-c-notification-drawer__list-item')
+      .eq(2)
+      .should('have.class', 'pf-m-warning');
+    cy.get('.pf-c-notification-drawer__list-item')
+      .last()
+      .should('have.class', 'pf-m-success');
+  });
+
+  it('Verify timestamp in list items', () => {
```
Comment (redallen) [inline packages/react-integration/cypress/integration/notificationdrawergroups.spec.ts:69]: Thanks for these tests!

File: packages/react-core/src/components/NotificationDrawer/NotificationDrawerGroup.tsx (hunk 1 above)
Comment (jcaianirh) [reply packages/react-core/src/components/NotificationDrawer/NotificationDrawerGroup.tsx:51]: Hmm I just rebased last night @redallen

Comment (redallen) [reply packages/react-core/src/components/NotificationDrawer/NotificationDrawerGroup.tsx:51]: `styles.notificationDrawerGroupToggleIcon` is the guy you want... I just checked and he's there on latest `v4`! You can try running `yarn build:generate` to regenerate react-styles.

**Comment by jcaianirh** [comment]
Thanks @redallen added https://github.com/patternfly/patternfly-react/issues/4143 to track that work item.

**Comment by jcaianirh** [comment]
@mcarrano , yes these behaviors are all baked into the component. This version of the notification drawer component includes what is used in openshift, but also a broader version including the read/unread status made available by props and optional child components.

File: packages/react-core/src/components/NotificationDrawer/NotificationDrawer.tsx [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import * as React from 'react';
+
+import styles from '@patternfly/react-styles/css/components/NotificationDrawer/notification-drawer';
+import { css } from '@patternfly/react-styles';
+
+export interface NotificationDrawerProps extends React.HTMLProps<HTMLDivElement> {
+  /**  Content rendered inside the Notification Drawer */
```
Comment (tlabaj) [inline packages/react-core/src/components/NotificationDrawer/NotificationDrawer.tsx:7]: Can you update this to be sentence case.  So it should be "notification drawer".

File: packages/react-core/src/components/NotificationDrawer/NotificationDrawerBody.tsx [hunk 4]
```diff
@@ -0,0 +1,21 @@
+import * as React from 'react';
+
+import styles from '@patternfly/react-styles/css/components/NotificationDrawer/notification-drawer';
+import { css } from '@patternfly/react-styles';
+
+export interface NotificationDrawerBodyProps extends React.HTMLProps<HTMLDivElement> {
+  /**  Content rendered inside the Body of the Notification Drawer */
```
Comment (tlabaj) [inline packages/react-core/src/components/NotificationDrawer/NotificationDrawerBody.tsx:7]: sentence case for these as well.

File: packages/react-core/src/components/NotificationDrawer/NotificationDrawerGroup.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+import AngleDownIcon from '@patternfly/react-icons/dist/js/icons/angle-down-icon';
+import AngleRightIcon from '@patternfly/react-icons/dist/js/icons/angle-right-icon';
+import { css } from '@patternfly/react-styles';
+import styles from '@patternfly/react-styles/css/components/NotificationDrawer/notification-drawer';
+
+import { Badge } from '../Badge';
+
+export interface NotificationDrawerGroupProps extends React.HTMLProps<HTMLElement> {
+  /**  Content rendered inside the Group */
```
Comment (tlabaj) [inline packages/react-core/src/components/NotificationDrawer/NotificationDrawerGroup.tsx:11]: Sentence case for these ones as well.

File: packages/react-core/src/components/NotificationDrawer/NotificationDrawerGroupList.tsx [hunk 6]
```diff
@@ -0,0 +1,21 @@
+import * as React from 'react';
+
+import { css } from '@patternfly/react-styles';
+import styles from '@patternfly/react-styles/css/components/NotificationDrawer/notification-drawer';
+
+export interface NotificationDrawerGroupListProps extends React.HTMLProps<HTMLDivElement> {
+  /**  Content rendered inside the Notification Drawer List Body */
```
Comment (tlabaj) [inline packages/react-core/src/components/NotificationDrawer/NotificationDrawerGroupList.tsx:7]: Sentence case.

File: packages/react-core/src/components/NotificationDrawer/NotificationDrawerList.tsx [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+import { css } from '@patternfly/react-styles';
+import styles from '@patternfly/react-styles/css/components/NotificationDrawer/notification-drawer';
+
+export interface NotificationDrawerListProps extends React.HTMLProps<HTMLUListElement> {
+  /**  Content rendered inside the Notification Drawer List Body */
+  children?: React.ReactNode;
+  /**  Additional classes added to the Notification Drawer List Body */
+  className?: string;
+  /**  Adds styling to the Notification Drawer List to indicate expand/hide state */
+  hidden?: boolean;
```
Comment (tlabaj) [inline packages/react-core/src/components/NotificationDrawer/NotificationDrawerList.tsx:12]: Can you change this name to isHidden for consistency please

File: packages/react-core/src/components/NotificationDrawer/NotificationDrawerListItem.tsx [hunk 8]
```diff
@@ -0,0 +1,48 @@
+import * as React from 'react';
+
+import { css, getModifier } from '@patternfly/react-styles';
+import styles from '@patternfly/react-styles/css/components/NotificationDrawer/notification-drawer';
+
+export interface NotificationDrawerListItemProps extends React.HTMLProps<HTMLLIElement> {
+  /**  Content rendered inside the List Item */
```
Comment (tlabaj) [inline packages/react-core/src/components/NotificationDrawer/NotificationDrawerListItem.tsx:7]: Sentence case.

File: packages/react-core/src/components/NotificationDrawer/NotificationDrawerHeader.tsx [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  className = '',
+  count,
+  title = 'Notifications',
+  ...props
+}: NotificationDrawerHeaderProps) => (
+  <div {...props} className={css(styles.notificationDrawerHeader, className)}>
+    <Text component={TextVariants.h1} className={css(styles.notificationDrawerHeaderTitle)}>
+      {title}
+    </Text>
+    {count && <span className={css(styles.notificationDrawerHeaderStatus)}>{`${count} unread`}</span>}
```
Comment (tlabaj) [inline packages/react-core/src/components/NotificationDrawer/NotificationDrawerHeader.tsx:30]: The unread string needs to me customizable for localization.  Can you add a prop for this please?

**Comment by jcaianirh** [comment]
Thanks @tlabaj. Pushed all changes from review comments.

**Comment by mcoker** [comment]
This looks fantastic @jcaianirh!!!

Some feedback:

1. The header action should probably have `position={DropdownPosition.right}` like the item action
Screen Shot 2020-04-30 at 4 35 26 PM
Screen Shot 2020-04-30 at 4 35 32 PM

2. When you expand a `.pf-c-notification-drawer__group` element, looks like that is applying a change to the `<svg>` arrow icon to show a down arrow. You shouldn't need to do that. It looks like you're missing the `.pf-m-expanded` class that applies to `.pf-c-notification-drawer__group` when it's expanded - that will tell the CSS to rotate the icon.

Here is the core code where you can see `.pf-c-notification-drawer__group.pf-m-expanded` and that we're using `fa-angle-right` as the icon and it's pointing down because the CSS rotates it.

Screen Shot 2020-04-30 at 4 43 53 PM


This will also fix the lack of border between the bottom of the expanded group and top of the next toggle.

This PR:

Screen Shot 2020-04-30 at 4 47 45 PM

Core:

Screen Shot 2020-04-30 at 4 49 03 PM

This will also fix the scroll behavior when the viewport height shrinks so that the expanded group height adjusts with the height of the viewport like below. Currently in this PR preview, the height of the notification drawer is static if you shrink the viewport shorter than the height of the notification drawer.

Apr-30-2020 16-53-58

**Comment by jcaianirh** [comment]
@mcoker thanks for the review! Made all the requested changes.

**Pull request merged**